_posts/2018-12-06-extended-memcpy.md

```diff
@@ -0,0 +1,153 @@
+---
+title: Extended memcpy in PMDK 1.5
+author: marcinslusarz
+layout: post
+identifier: extended_memcpy
+---
+
+In PMDK 1.5 we added new APIs for bulk persistent memory modifications.
+In short we did this to:
+ - give applications ability to perform low-level performance optimizations
+ - clean up the unfortunate naming
+
+In order to understand what exactly and why we did that, let's review the API
+as exposed by PMDK 1.4 and earlier. In PMDK 1.4 we had these functions:
+
+{% highlight C %}
+void *pmem_memmove_persist(void *pmemdest, const void *src, size_t len);
+void *pmem_memcpy_persist (void *pmemdest, const void *src, size_t len);
+void *pmem_memset_persist (void *pmemdest, int c, size_t len);
+
+void *pmem_memmove_nodrain(void *pmemdest, const void *src, size_t len);
+void *pmem_memcpy_nodrain (void *pmemdest, const void *src, size_t len);
+void *pmem_memset_nodrain (void *pmemdest, int c, size_t len);
+
+void *pmemobj_memcpy_persist(PMEMobjpool *pop, void *pmemdest, const void *src, size_t len);
+void *pmemobj_memset_persist(PMEMobjpool *pop, void *pmemdest, int c, size_t len);
+{% endhighlight %}
+
+As you can see there are two variants of these APIs - with `_persist` and
+`_nodrain` suffix. Both variants modify pmemdest argument to do what their
+libc's equivalent do and make sure that the copied data are flushed.
+The only difference is that the `_persist` variants also wait for the flush to
+finish. As I mentioned earlier, there's one unfortunate aspect of this naming -
+it doesn't follow the scheme used by other libpmem functions, where
+`pmem_flush` flushes the data, `pmem_drain` waits for all flushes to finish
+and `pmem_persist` does both. In line with this scheme `_nodrain` functions
+should have used `_flush` suffix.
+
+For libpmemobj functions there's no way to opt-out of drains and there's no
+memmove. Functionality-wise they are wrappers around libpmem functions with
+additions required by libpmemobj (replication and non-pmem safety).
+
+The bigger problem than naming is with what they do - they try to guess
+the optimal strategy to update persistent memory using only their parameters.
+The logic is pretty simple - if modification is smaller than 256 bytes
+(configurable using `PMEM_MOVNT_THRESHOLD` environment variable) they use
+normal `mov` instrunctions followed by `pmem_flush`, but for modifications
+256 bytes and longer they use non-temporal stores. Those instructions on
+x86_64 have 2 properties - they bypass CPU caches (so that pmem_flush is not
+needed) and treat destination memory as the write-combining type.
+The last property means that if the destination memory is not in the cache,
+the CPU doesn't have to fetch the full cache lines, just to flush them a moment
+later. This is important not because data is not stored in the cache, but
+because there's no fetch of previous data.
+This means that application can update the same piece of memory multiple times
```

Comment:
`can update the same cacheline multiple times?`

"Piece of memory" is ambiguous.